samples/matter/bridge/src/bridge_manager.cpp

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "bridge_manager.h"
+
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/util/generic-callbacks.h>
+#include <lib/support/Span.h>
+#include <lib/support/ZclString.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);
+
+using namespace ::chip;
+using namespace ::chip::app;
+
+BridgeManager BridgeManager::sBridgeManager;
+
+void BridgeManager::Init()
+{
+	/* The first dynamic endpoint is the last fixed endpoint + 1. */
+	mFirstDynamicEndpointId = static_cast<chip::EndpointId>(
+		static_cast<int>(emberAfEndpointFromIndex(static_cast<uint16_t>(emberAfFixedEndpointCount() - 1))) + 1);
+
+	mCurrentDynamicEndpointId = mFirstDynamicEndpointId;
+
+	/* Disable the placeholder endpoint */
+	emberAfEndpointEnableDisable(emberAfEndpointFromIndex(static_cast<uint16_t>(emberAfFixedEndpointCount() - 1)),
+				     false);
+}
+
+CHIP_ERROR BridgeManager::AddBridgedDevice(BridgedDevice::DeviceType bridgedDeviceType, char *nodeLabel)
+{
+	BridgedDevice *device;
+
+	char *label = nullptr;
+	if (nodeLabel && (strlen(nodeLabel) >= BridgedDevice::kNodeLabelSize)) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (nodeLabel) {
+		label = nodeLabel;
+	}
+
+	switch (bridgedDeviceType) {
+#if CONFIG_BRIDGE_ONOFF_LIGHT_BRIDGED_DEVICE
+	case BridgedDevice::DeviceType::OnOffLight: {
+		LOG_INF("Adding OnOff Light bridged device");
+		device = reinterpret_cast<BridgedDevice *>(Platform::New<OnOffLightDevice>(label));
+		break;
+	}
+#endif /* CONFIG_BRIDGE_ONOFF_LIGHT_BRIDGED_DEVICE */
+#if CONFIG_BRIDGE_TEMPERATURE_SENSOR_BRIDGED_DEVICE
+	case BridgedDevice::DeviceType::TemperatureSensor: {
+		LOG_INF("Adding TemperatureSensor bridged device");
+		device = reinterpret_cast<BridgedDevice *>(Platform::New<TemperatureSensorDevice>(label));
+		break;
+	}
+#endif /* CONFIG_BRIDGE_TEMPERATURE_SENSOR_BRIDGED_DEVICE */
+#if CONFIG_BRIDGE_HUMIDITY_SENSOR_BRIDGED_DEVICE
+	case BridgedDevice::DeviceType::HumiditySensor: {
+		LOG_INF("Adding HumiditySensor bridged device");
+		device = reinterpret_cast<BridgedDevice *>(Platform::New<HumiditySensorDevice>(label));
+		break;
+	}
+#endif /* CONFIG_BRIDGE_HUMIDITY_SENSOR_BRIDGED_DEVICE */
+	default:
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (!device) {
+		return CHIP_ERROR_NO_MEMORY;
+	}
+
+	return AddDevice(device);
+}
+
+CHIP_ERROR BridgeManager::RemoveBridgedDevice(uint16_t endpoint)
+{
+	uint8_t index = 0;
+
+	while (index < kMaxBridgedDevices) {
+		if (nullptr != mBridgedDevices[index]) {
+			if (mBridgedDevices[index]->GetEndpointId() == endpoint) {
+				LOG_INF("Removed dynamic endpoint %d (index=%d)", endpoint, index);
+				/* Free dynamically allocated memory */
+				emberAfClearDynamicEndpoint(index);
+				Platform::Delete(mBridgedDevices[index]);
```

Comment:
You must add virtual d-tor `BridgedDevice` no to leak memory.

Reply:
Right, fixed